### PR TITLE
Handle inherited messages correctly

### LINKED ISSUE #95: NServiceBus.Testing V6 treats message inheritence differently
When upgrading to V6, I'm noticing a difference in the way the testing framework locates the appropriate message handler from prior versions. It seems much more strict, and doesn't seem to reflect how the framework would work at runtime.

given messages:

```cs
    public abstract class BaseCommand
    {
    }
    public class Command : BaseCommand
    {
    }
    public interface IMyEvent
    {
    }
    public class MyEvent : IMyEvent
    {
    }
```

this handler

```cs
    public class TestHandler : IHandleMessages<BaseCommand>, IHandleMessages<IMyEvent>
    {
        public Task Handle(BaseCommand message, IMessageHandlerContext context)
        {
            return context.Publish<IMyEvent>();
        }

        public Task Handle(IMyEvent message, IMessageHandlerContext context)
        {
            throw new NotImplementedException("can't handle my event");
        }
    }

```

this test
```cs
    [TestFixture]
    public class TextFixture
    {
        [Test]
        public void HandlesCommand()
        {

            Test.Handler<TestHandler>()
                .ExpectPublish<IMyEvent>()
                .OnMessage(new Command());
         //this test never actually calls the handler, and doesn't fail either :(
        }

        [Test]
        public void HandlesEvent()
        {
            Assert.Throws<NotImplementedException>(() => Test.Handler<TestHandler>()
                .OnMessage(new MyEvent()));
         //this test fails, Handle is never called
        }

    }

```

This scenario works in prior version of the testing framework

### PR DESCRIPTION
**Pull request by timbussmann**
fixes #95 

the testing framework isn't able to handle message hierarchies properly and also invokes only one handle method per message.

@Particular/nservicebus-maintainers please review

### CODE REVIEW

File: src/NServiceBus.Testing/TypeExtensions.cs [hunk 1]
```diff
@@ -1,39 +1,48 @@
...
+
+                yield return Expression.Lambda<Func<object, object, IMessageHandlerContext, Task>>(body, target, messageParam, contextParam).Compile();
+            }
          }
+
+         public static async Task InvokeSerially(this IEnumerable<Func<object, object, IMessageHandlerContext, Task>> invokers, object instance, object message, IMessageHandlerContext context)
+         {
+            foreach (var invocation in invokers)
+            {
+                await invocation(instance, message, context);
```
Comment (andreasohlund) [inline src/NServiceBus.Testing/TypeExtensions.cs]: No .ConfigureAwait(false) since this is a test fx right?

Comment (danielmarbach) [inline src/NServiceBus.Testing/TypeExtensions.cs]: I think it is actually needed here. Test Fxs can have synchronization contexts as well.

Comment (timbussmann) [inline src/NServiceBus.Testing/TypeExtensions.cs]: needed is such a strong word ;) but yeah I can add it.

**Comment by timbussmann** [comment]
I added the `Particular.CodeRules` package to automate the ConfigureAwait checks.

**Pull request merged**